[HUDI-8474] Adding upsert prepped api to write client for optimized flow for metadata table

nsivabalan opened the pull request:
### Change Logs

Adding upsert prepped api to write client for optimized flow for metadata table.
this is stacked on top of https://github.com/apache/hudi/pull/13005 

This new write client api is not yet integrated anywhere. 

We are going to raise lot of smaller patches as part of https://issues.apache.org/jira/browse/HUDI-9281. Goal is to keep producing small patches that can land in isolation w/o being blocked on very large patch. 
for overall design, feel free to check out https://github.com/apache/hudi/pull/13236 

### Impact

_Describe any public API or user-facing feature change or any performance impact._

### Risk level (write none, low medium or high below)

_If medium or high, explain what verification was done to mitigate the risks._

### Documentation Update

_Describe any necessary documentation update if there is any new feature, config, or user-facing change. If not, put "none"._

- _The config description must be updated if new configs are added or the default value of the configs are changed_
- _Any new feature or user-facing change requires updating the Hudi website. Please create a Jira ticket, attach the
  ticket number here and follow the [instruction](https://hudi.apache.org/contribute/developer-setup#website) to make
  changes to the website._

### Contributor's checklist

- [ ] Read through [contributor's guide](https://hudi.apache.org/contribute/how-to-contribute)
- [ ] Change Logs and Impact were stated clearly
- [ ] Adequate tests were added if applicable
- [ ] CI passed

the-other-tim-brown commented on hudi-client/hudi-spark-client/src/test/java/org/apache/hudi/table/action/commit/TestSparkMetadataTableUpsertPartitioner.java:
```diff
@@ -0,0 +1,97 @@
+  private static final String RANDOM_INSTANT_TIME = "100000";
+
+  @Test
+  public void testUpsertPartitioner() throws Exception {
```
nitpick: `throws Exception` is not required

nsivabalan commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/SparkRDDWriteClient.java:
```diff
@@ -148,6 +152,24 @@ public JavaRDD<WriteStatus> upsertPreppedRecords(JavaRDD<HoodieRecord<T>> preppe
+        initTable(WriteOperationType.UPSERT_PREPPED, Option.ofNullable(instantTime));
+    table.validateUpsertSchema();
+    if (!(partitionFileIdPairsOpt.isPresent() && partitionFileIdPairsOpt.get().get(0).getKey().equals(FILES.getPartitionPath()))) {
+      // we do not want to call prewrite more than once for the same instant writing to metadata table twice.
```
Alternate idea I had was to maintain a set<String> i.e. Set<Instant time> for which pre-write is already invoked. But since this is only applicable for MDT, did not want to make this generic or all cases. If we might have use-cases for data table, where we might wanna do similar things 
i.e. 
```
t1 = writeClient.startCommit()
writeClient.upsert(batch1, t1)
writeClient.upsert(batch2, t1)
writeClient.commit 
```

then may be, we can make a justification for add a hashset to writeClient class.

the-other-tim-brown replied:
Now that there is a separate write client for the mdt writer, is it worth revisiting?

nsivabalan replied:
yes, I have pushed an update to fix this.

nsivabalan commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/SparkRDDWriteClient.java:
```diff
@@ -148,6 +152,24 @@ public JavaRDD<WriteStatus> upsertPreppedRecords(JavaRDD<HoodieRecord<T>> preppe
+    HoodieTable<T, HoodieData<HoodieRecord<T>>, HoodieData<HoodieKey>, HoodieData<WriteStatus>> table =
+        initTable(WriteOperationType.UPSERT_PREPPED, Option.ofNullable(instantTime));
+    table.validateUpsertSchema();
+    if (!(partitionFileIdPairsOpt.isPresent() && partitionFileIdPairsOpt.get().get(0).getKey().equals(FILES.getPartitionPath()))) {
```
👍

nsivabalan commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/action/commit/SparkMetadataTableUpsertCommitActionExecutor.java:
```diff
@@ -0,0 +1,100 @@
+  private static final Logger LOG = LoggerFactory.getLogger(SparkMetadataTableUpsertCommitActionExecutor.class);
+
+  private static final HashMap<String, WorkloadStat> EMPTY_MAP = new HashMap<>();
+  private static final WorkloadStat PLACEHOLDER_GLOBAL_STAT = new WorkloadStat();
```
yes. atleast for MDT writes, we do not modify entire workload.

nsivabalan commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/HoodieSparkMergeOnReadTable.java:
```diff
@@ -126,6 +128,14 @@ public HoodieWriteMetadata<HoodieData<WriteStatus>> upsertPrepped(HoodieEngineCo
   }
 
+  @Override
+  public HoodieWriteMetadata<HoodieData<WriteStatus>> upsertPrepped(HoodieEngineContext context, String instantTime,
```
Actually I am adding a new class called SparkRDDMetadataWriteClient and moving 
```
upsertPreppedRecords(JavaRDD<HoodieRecord<T>> preppedRecords, String instantTime, Option<List<Pair<String, String>>> partitionFileIdPairsOpt) {
```

in there.

nsivabalan commented:
hey @the-other-tim-brown : I have fixed the source code. Added separate write client for metadata table and HoodieTable as well. Will be adding tests by tomorrow. But feel free to review in the mean time if you are interested.

the-other-tim-brown commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/BaseHoodieWriteClient.java:
```diff
@@ -428,6 +431,19 @@ public void bootstrap(Option<Map<String, String>> extraMetadata) {
+   * @param instantTime Instant time of the commit
+   * @return Collection of WriteStatus to inspect errors and counts
+   */
+  public O upsertPreppedRecords(I preppedRecords, final String instantTime, Option<List<Pair<String, String>>> partitionFileIdPairsOpt) {
```
Is there a way to avoid adding this to the base client now since it is specific to spark mdt writer?

nsivabalan replied:
yes.

the-other-tim-brown commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/HoodieSparkTable.java:
```diff
@@ -75,6 +75,8 @@ public static <T> HoodieSparkTable<T> create(HoodieWriteConfig config,
         hoodieSparkTable = new HoodieSparkCopyOnWriteTable<>(config, context, metaClient);
         break;
       case MERGE_ON_READ:
+        // to fix.
+        // if metadata table -> new HoodieSparkMergeOnReadMetadataTable<>(config, context, metaClient);
```
Does this need to happen in this PR?

the-other-tim-brown commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/SparkRDDMetadataWriteClient.java:
```diff
@@ -0,0 +1,71 @@
+  public JavaRDD<WriteStatus> upsertPreppedRecords(JavaRDD<HoodieRecord<T>> preppedRecords, String instantTime, Option<List<Pair<String, String>>> partitionFileIdPairsOpt) {
+    ValidationUtils.checkArgument(isMetadataTable, "This version of upsert prepped can only be invoked for metadata table");
+    HoodieTable<T, HoodieData<HoodieRecord<T>>, HoodieData<HoodieKey>, HoodieData<WriteStatus>> table =
+        initTable(WriteOperationType.UPSERT_PREPPED, Option.ofNullable(instantTime));
```
Do you need to override the init table in this class to return the MDT version of spark table?

nsivabalan replied:
w/ my latest commit, createHoodieTable will take care of returning the right table(data table or metadata table)

nsivabalan commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/action/commit/SparkMetadataTableUpsertCommitActionExecutor.java:
```diff
@@ -0,0 +1,100 @@
+    // with streaming writes support, we might write to metadata table multiple times for the same instant times.
+    // ie. writeClient.startCommit(t1), writeClient.upsert(batch1, t1), writeClient.upsert(batch2, t1), writeClient.commit(t1, ...)
+    // So, here we are generating inflight file only in the last known writes, which we know will only have FILES partition.
+    if (mdtPartitionPathFileGroupIdList.size() == 1 && mdtPartitionPathFileGroupIdList.get(0).getKey().equals(FILES.getPartitionPath())) {
```
only other alternative that I thought about is:
w/ latest update to this patch, we can determine elegantly in SparkRDDMetadataWriteClient whether its a first call or 2nd call. So, while calling into HoodieTable.upsertPrepped().. we need to add an additional arg to denote whether its first call or 2nd call. which in turn will get passed to SparkMetadataTableUpsertCommitActionExecutor constructor. 

which I felt is too much of an intrusive change just to achieve this.

nsivabalan commented:
@the-other-tim-brown : addressed all feedback

nsivabalan commented:
@the-other-tim-brown : addressed all feedback

the-other-tim-brown commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/BaseHoodieWriteClient.java:
```diff
@@ -142,6 +143,7 @@ public abstract class BaseHoodieWriteClient<T, I, K, O> extends BaseHoodieClient
   protected Set<String> pendingInflightAndRequestedInstants = Collections.emptySet();
 
   protected BaseHoodieTableServiceClient<?, ?, O> tableServiceClient;
+  protected boolean isMetadataTable;
```
This looks like it is unused now, can it be removed?

the-other-tim-brown commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java:
```diff
@@ -1463,10 +1468,11 @@ protected abstract void bulkCommit(
    * The record is tagged with respective file slice's location based on its record key.
    */
-  protected HoodieData<HoodieRecord> prepRecords(Map<String, HoodieData<HoodieRecord>> partitionRecordsMap) {
+  protected Pair<HoodieData<HoodieRecord>, List<Pair<String, String>>> prepRecords(Map<String, HoodieData<HoodieRecord>> partitionRecordsMap) {
```
Can you update the javadocs so it is clear what is being returned? It may also be worthwhile to introduce a POJO so it is more clear what the `Pair<String, String>` represents

the-other-tim-brown commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java:
```diff
@@ -1481,18 +1487,21 @@ protected HoodieData<HoodieRecord> prepRecords(Map<String, HoodieData<HoodieReco
           r.unseal();
           r.setCurrentLocation(new HoodieRecordLocation(slice.getBaseInstantTime(), slice.getFileId()));
           r.seal();
+          mappedFileIds.add(slice.getFileId());
```
The `mappedFileIds` set will be serialized to the executor in the spark context. It will not be reported back to the driver so I don't think this implementation will work.

nsivabalan replied:
good catch. I did fix it in my later patch, but missed to update here. 

https://github.com/apache/hudi/pull/13312/files#r2112561019 
you can check it out here. 

there is a known limitation to this. we will be spinning up a spark task while writing for all file groups for the mdt partitions we are touching, since w/o triggering the action on the records, we don't really know which among the file groups actually have records to be written to. 

but since we do not want to trigger any action, we are returning every file group in the partiions we touch from here.

nsivabalan replied:
will fix this patch accordingly.

the-other-tim-brown commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/SparkRDDMetadataWriteClient.java:
```diff
@@ -0,0 +1,98 @@
+    HoodieTable<T, HoodieData<HoodieRecord<T>>, HoodieData<HoodieKey>, HoodieData<WriteStatus>> table =
+        initTable(WriteOperationType.UPSERT_PREPPED, Option.ofNullable(instantTime));
+    table.validateUpsertSchema();
+    boolean initialCall = preWriteCompletedInstants.contains(instantTime);
+    if (!initialCall) {
```
I think the naming here is off, it seems like we want to perform `preWrite` on the initial invocation, right?

nsivabalan commented:
@the-other-tim-brown : addressed the feedback.

the-other-tim-brown commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java:
```diff
@@ -1408,6 +1411,8 @@ protected void commitInternal(String instantTime, Map<String, HoodieData<HoodieR
       bulkInsertAndCommit(writeClient, instantTime, preppedRecordInputs, bulkInsertPartitioner);
     } else {
       engineContext.setJobStatus(this.getClass().getSimpleName(), String.format("Upserting at %s into metadata table %s", instantTime, metadataWriteConfig.getTableName()));
+      // to do: fix the last argument is required so that we can support streaming writes to metadata table.
+      // Option.of(partitionFileIdPairs)
```
if this is handled in a later PR, my suggestion is to just remove the TODO from here

the-other-tim-brown commented on hudi-client/hudi-spark-client/src/test/java/org/apache/hudi/client/TestSparkRDDMetadataWriteClient.java:
```diff
@@ -0,0 +1,414 @@
+      HoodieActiveTimeline reloadedMdtActiveTimeline = metadataMetaClient.reloadActiveTimeline();
+      assertEquals(reloadedMdtActiveTimeline.filterCompletedInstants().getInstants().stream().count(), 3); // files, rli instantiaton and 1 write to data table.
+      String finalCommitTimeOfInterest = commitTimeOfInterest;
+      assertTrue(reloadedMdtActiveTimeline.filterInflightsAndRequested().getInstants().stream().anyMatch(instant -> instant.requestedTime().equals(finalCommitTimeOfInterest)));
```
If there are two `.inflight` files for a commit, will that throw an exception somewhere? Wondering if we want to validate this by listing the directory manually and inspecting it.

the-other-tim-brown commented on hudi-client/hudi-spark-client/src/test/java/org/apache/hudi/client/TestSparkRDDMetadataWriteClient.java:
```diff
@@ -0,0 +1,414 @@
+      fileSlices =
+          HoodieTableMetadataUtil.getPartitionLatestFileSlices(metadataMetaClient, Option.ofNullable(fsView), MetadataPartitionType.RECORD_INDEX.getPartitionPath());
+      mdtPartitionsFileIdMapping.put(MetadataPartitionType.RECORD_INDEX.getPartitionPath(), fileSlices.stream().map(fileSlice -> fileSlice.getFileId()).collect(Collectors.toList()));
+      fileSlices.stream().forEach(fileSlice -> {
```
instead of collecting in the step above, you can just use `forEach` directly

yihua commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/SparkRDDMetadataWriteClient.java:
```diff
@@ -0,0 +1,98 @@
+public class SparkRDDMetadataWriteClient<T> extends SparkRDDWriteClient<T> {
+
+  // tracks the instants for which preWrite has been invoked.
+  private final Set<String> preWriteCompletedInstants = new HashSet<>();
```
If the assumption is that `SparkRDDMetadataWriteClient` is for one data table instant / action, it would be better to maintain a single instant here for validation, and throw an exception if `#upsertPreppedRecords` and `#commit` is called on a different instant.

yihua commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/SparkRDDMetadataWriteClient.java:
```diff
@@ -0,0 +1,98 @@
+    HoodieTable<T, HoodieData<HoodieRecord<T>>, HoodieData<HoodieKey>, HoodieData<WriteStatus>> table =
+        initTable(WriteOperationType.UPSERT_PREPPED, Option.ofNullable(instantTime));
+    table.validateUpsertSchema();
+    boolean initialCall = !preWriteCompletedInstants.contains(instantTime);
+    if (initialCall) {
+      // we do not want to call prewrite more than once for the same instant, since we could be writing to metadata table more than once w/ streaming writes.
+      preWrite(instantTime, WriteOperationType.UPSERT_PREPPED, table.getMetaClient());
+      preWriteCompletedInstants.add(instantTime);
+    }
```
This can be tighter by checking exactly twice invocation.

yihua commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/SparkRDDMetadataWriteClient.java:
```diff
@@ -0,0 +1,98 @@
+  public boolean commit(String instantTime, JavaRDD<WriteStatus> writeStatuses, Option<Map<String, String>> extraMetadata,
+                        String commitActionType, Map<String, List<String>> partitionToReplacedFileIds,
+                        Option<BiConsumer<HoodieTableMetaClient, HoodieCommitMetadata>> extraPreCommitFunc) {
+    preWriteCompletedInstants.remove(instantTime);
```
Similar here on validation.

yihua commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/SparkRDDMetadataWriteClient.java:
```diff
@@ -0,0 +1,98 @@
+   * Complete changes performed at the given instantTime marker with specified action.
+   */
+  @Override
+  public boolean commit(String instantTime, JavaRDD<WriteStatus> writeStatuses, Option<Map<String, String>> extraMetadata,
```
nit: one option is to let this metadata write client track multiple `JavaRDD<WriteStats>` from `#upsertPreppedRecords` and does the merging before passing to `#commit`.  However, the reason we don't do this is because of how streaming write DAG is constructed.

yihua replied:
Now I understand this: `JavaRDD<WriteStatus> writeStatuses` needs a better naming for readability.  Or override the `commitStats`.

danny0405 commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java:
```diff
@@ -1370,7 +1371,9 @@ public void close() throws Exception {
-    HoodieData<HoodieRecord> preppedRecords = prepRecords(partitionRecordsMap);
+    Pair<HoodieData<HoodieRecord>, List<HoodieFileGroupId>> result = prepRecords(partitionRecordsMap);
+    HoodieData<HoodieRecord> preppedRecords = result.getKey();
+    List<HoodieFileGroupId> hoodieFileGroupIdList = result.getValue();
```
`hoodieFileGroupIdList` is not used in anywhere.

danny0405 commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/SparkRDDMetadataWriteClient.java:
```diff
@@ -0,0 +1,98 @@
+                                     Option<EmbeddedTimelineService> timelineService) {
+    super(context, writeConfig, timelineService);
+  }
+
```
personally I don't very much like over-protection, if we are sure only `upsertPreppedRecords` would be invoked for MDT.

danny0405 commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/action/commit/SparkMetadataTableUpsertCommitActionExecutor.java:
```diff
@@ -0,0 +1,103 @@
+    // with streaming writes support, we might write to metadata table multiple times for the same instant times.
+    // ie. writeClient.startCommit(t1), writeClient.upsert(batch1, t1), writeClient.upsert(batch2, t1), writeClient.commit(t1, ...)
+    // So, here we are generating inflight file only in the last known writes, which we know will only have FILES partition.
+    if (initialCall) {
```
should we just do nothing because the `profile` is always empty.

nsivabalan commented:
@yihua @danny0405 : addressed all feedback. PTAL

the-other-tim-brown reviewed (approved):
Looks good to me, if the FileGroupReader for metadata table lands soon, lets swap that in for the testing

nsivabalan commented:
@the-other-tim-brown : sure.

yihua commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/SparkRDDMetadataWriteClient.java:
```diff
@@ -0,0 +1,90 @@
+      ValidationUtils.checkArgument(firstInstantOpt.get().equals(instantTime), "Upsert Prepped invoked for metadata table using same write client instance "
+          + " for two different instant times " + firstInstantOpt.get() + " and " + instantTime);
+    }
+    ValidationUtils.checkArgument(invocationCounts <= 2, "Upsert Prepped invoked more then twice for the same instant time with metadata write client "
```
To clarify, if no index is written through the streaming DAG (e.g., only files partition is enabled), `upsertPreppedRecords`, the `invocationCounts` is one?

yihua reviewed (approved):
LGTM
